### PR TITLE
Fix windows support for `build_executable.jl`. fix #9973

### LINKED ISSUE #9973: build_executable.jl windows support
I am trying to build a standalone executable on windows and I'm running into some problems.
I have added gcc as stated in the script.

First of I ran into a problem with a missing string escape when windows paths is written to a userimg file in `emit_userimgjl`. It causes the userimg to contain an `include("folder\folder\file.jl")` with only single `\`. By inserting include(\"$(escape_string(script_file))\") on line: https://github.com/JuliaLang/julia/blob/26ae20698cf0fc57ae6fc60ba42140884e71f558/contrib/build_executable.jl#L224 the path to the script is correct on windows.

After overcoming this problem I get this error:
```
INFO: Linking sys.dll
INFO: System image successfully built at C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\libeps_to_pdf.dll
WARNING: Building sys.dll on Windows against LLVM < 3.5.0 can cause incorrect backtraces! Delete generated sys.dll to avoid these
problems
INFO: To run Julia with this image loaded, run: julia -J C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\libeps_to_pdf.ji

running: gcc -g -Wl,--no-as-needed -IC:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia -IC:\Users\Hoegh\src -IC:\Users\Hoegh\src/
upport -IC:\Users\Hoegh\usr/include C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c -o C:\Users\Hoegh\Julia-0.4.0-dev_
9_1\bin\eps_to_pdf.exe -Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -LC:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -ljulia -leps
to_pdf
ERROR: LoadError: could not spawn `gcc -g -Wl,--no-as-needed '-IC:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia' '-IC:\Users\Ho
gh\src' '-IC:\Users\Hoegh\src/support' '-IC:\Users\Hoegh\usr/include' 'C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c
 -o 'C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin\eps_to_pdf.exe' '-Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin' '-LC:\Users\Hoe
h\Julia-0.4.0-dev_19_1\bin' -ljulia -leps_to_pdf`: no such file or directory (ENOENT)
 in _jl_spawn at process.jl:216 (repeats 2 times)
while loading C:\Users\Hoegh\Julia-0.4.0-dev_19_1\share\julia\build_executable.jl, in expression starting on line 247
```
When I run the command manually i get:
```
C:\Users\Hoegh\Julia-0.4.0-dev_19_1>gcc -g -Wl,--no-as-needed -IC:\Users\Hoegh\Julia-0.4.0-dev_19_
1\include\julia -IC:\Users\Hoegh\src -IC:\Users\Hoegh\src/support -IC:\Users\Hoegh\usr/include C:\
Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c -o C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin
\eps_to_pdf.exe -Wl,-rpath,C:\Users\Hoegh\Julia-0.4.0-dev_19_1\bin -LC:\Users\Hoegh\Julia-0.4.0-de
v_19_1\bin -ljulia -leps_to_pdf
In file included from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/uv.h:59:0,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/ios.h:5,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/libsupport.h:12,
                 from C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/julia.h:10,
                 from C:\Users\Hoegh\AppData\Local\Temp\jul9D3B.tmp\start_func.c:1:
C:\Users\Hoegh\Julia-0.4.0-dev_19_1\include\julia/uv-win.h:241:3: error: unknown type name 'SRWLOCK'
```
The file I try to compile looks like:
```
function main()
    println("test")
end
```
@few do you have any idea what's wrong?

### PR DESCRIPTION
**Pull request by dhoegh**
The changes made to the script allow for use on windows.

I have changed how `gcc` is found, the setup for the compiler and the changes in the copying/moving files. The things I would like to get thorough review is:
- That I have disabled the fix rpath on windows, https://github.com/dhoegh/julia/blob/b76f234f777cb37f4686af11200e9ec928622908/contrib/build_executable.jl#L156-L168. The results still works also if the containing folder is moved, I do not know if this is correct please enlighten me.
-  Changes in `find_patchelf` function. I changed it due to it did not return `nothing` when `patchelf` is not installed, which I think is not the intention.

It would also be great if someone could test it on windows, if possible windows xp due to https://github.com/dhoegh/julia/blob/b76f234f777cb37f4686af11200e9ec928622908/contrib/build_executable.jl#L114 which I have not tested but brought up in #9973.

@few could you review my changes?

### CODE REVIEW

File: contrib/build_executable.jl [hunk 1]
```diff
@@ -109,37 +108,62 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
     run(`$(julia) $(build_sysimg) $(sys.buildfile) native $(userimgjl) --force`)
     println()
 
+    gcc = find_system_gcc()
+    # This argument is needed for the gcc, see issue #9973
+    win_arg = @windows ? "-D_WIN32_WINNT=0x0600" : ""
+    @windowsxp_only win_arg = "-D_WINVER=0x0502"
```
Comment (tkelman) [inline contrib/build_executable.jl]: could you not use this same 0x0502 setting on newer windows?

File: contrib/build_executable.jl [hunk 2]
```diff
@@ -219,12 +236,30 @@ function emit_cmain(cfile, exename, relocation)
...
+        winrpmgcc = joinpath(WinRPM.installdir,"usr","$(Sys.ARCH)-w64-mingw32",
+            "sys-root","mingw","bin","gcc.exe")
+        if success(`$winrpmgcc --version`)
+            return winrpmgcc
+        end
+    end
+
+    # See if `gcc` exists
+    try
+        if success(@windows ? `cmd /c gcc -v` : `gcc -v`)
```
Comment (tkelman) [inline contrib/build_executable.jl]: Please remove this. I don't want to deal with picking up random other incompatible gcc's that come from god knows where else and end up on people's paths.

File: contrib/build_executable.jl (hunk 1 above)
Comment (dhoegh) [inline contrib/build_executable.jl]: So change "-D_WIN32_WINNT=0x0600" to "-D_WIN32_WINNT=0x0502"?

Comment (tkelman) [inline contrib/build_executable.jl]: Yes, try that, use it if it works. The important part isn't trying to build things on an XP machine, as much as making sure the results can run there.

File: contrib/build_executable.jl [hunk 3]
```diff
@@ -109,37 +108,62 @@ function build_executable(exename, script_file, targetdir=nothing; force=false)
...
+    gcc = find_system_gcc()
+    # This argument is needed for the gcc, see issue #9973
+    win_arg = @windows ? "-D_WIN32_WINNT=0x0600" : ""
+    @windowsxp_only win_arg = "-D_WINVER=0x0502"
+    incs = get_includes()
+    @windows_only begin
+        if contains(gcc, "WinRPM")
+            # This should not bee necessary, it is done due to WinRPM's gcc's
+            # include paths is not correct see WinRPM.jl issue #38
+            old_path = ENV["path"]
```
Comment (tkelman) [inline contrib/build_executable.jl]: Don't modify path. Make a copy of `ENV`, something like `ENV2 = copy(ENV)`. Then modify `ENV2["PATH"]` (please refer to that variable in all caps) and execute commands with ```setenv(`gcc blah`, ENV2)```

File: contrib/build_executable.jl (hunk 2 above)
Comment (tkelman) [inline contrib/build_executable.jl]: So this `try` block checking whichever `gcc` is on the path should be `@unix_only`.

**Comment by dhoegh** [comment]
Thank you @tkelman for sharing you expertise, I have implemented your corrections.

**Comment by dhoegh** [comment]
Bump. @few do you have time for reviewing the pull?

**Comment by few** [comment]
I took a quick look. I assume that you made sure it still works under linux. 
1) You made the patchelf magic unix_only. I guess this is because windows loads dependent shared libs from the directory a shared lib is in?
2) The error message for the missing gcc might look at bit strange on linux.

Other than that, no objections.

**Comment by dhoegh** [comment]
Same on me I have not tested it on Linux/mac due to my Linux machine is broken. Could you test it, if not I need to come around to a Linux? 

It would be nice if it could be tested on Travis/Appveyor, but I think this would overload the the services. Maybe this indicates the script belongs in a separate repository.
1) yes
2) The error message is windows dependant so it should not print on Linux.

**Comment by dhoegh** [comment]
I have now tested the `build_executable.jl` on Ubuntu 14.10 and found that if `targetdir` where not set then the script did not work on Ubuntu or Windows.  With the latest changes this is fixed.

From my point of this is good to merge.

**Comment by tkelman** [comment]
If you're done and happy with this for now, LGTM.

**Pull request merged**